### PR TITLE
Add support for selective erase operations

### LINKED ISSUE #14029: Add support for selective erase operations
# Description of the new feature/enhancement

I'd like to add support for the selective erase escape sequences: [`DECSED`](https://vt100.net/docs/vt510-rm/DECSED.html), [`DECSEL`](https://vt100.net/docs/vt510-rm/DECSEL.html), and [`DECSCA`](https://vt100.net/docs/vt510-rm/DECSCA.html). They provide a way of marking certain areas of the screen as "protected", and then you can easily erase all but the protected areas.

They were typically used in data entry applications, where you'd have a form with static content like the labels that you'd want protected. Then when you needed to reset the form, you could just clear the whole screen (with `DECSED`), and you wouldn't have to redraw those protected areas of the UI.

I could also see them being useful in certain gaming situations, where you might have a static background (e.g. the maze walls in Pac-Man), which you'd want to protect. Then you could easily clear out all the non-static sprites with `DECSED`, without losing the static background.

However, the main reason I wanted to add them was as a precursor to doing the rectangular area operations, since once of those ([`DECSERA`](https://vt100.net/docs/vt510-rm/DECSERA.html)) is dependent on protected attributes.

# Proposed technical implementation details (optional)

We need another bit in the `ExtendedAttributes` enum to hold the protected status of each cell, an operation to enable and disable that bit in the active attributes (this is `DECSCA`), and variants of the `EraseInDisplay` and `EraseInLine` methods that only erase the cells that don't have the protected bit set.

### PR DESCRIPTION
**Pull request by j4james**
This PR adds support for the selective erase escape sequences: `DECSED`,
`DECSEL`, and `DECSCA`. They provide a way of marking certain areas of
the screen as "protected", so you can erase the content everywhere else
without affecting those protected areas.

This adds another bit in the `CharacterAttributes` enum to track the
protected status of each cell, and an operation triggered by the
`DECSCA` sequence which can toggle that bit in the active attributes.

There there are two new erase operations triggered by the `DECSED` and
`DECSEL` sequences, which work similar to the existing `ED` (erase in
display) and `EL` (erase in line) operations, but which only apply to
unprotected cells.

I've also updated the `DECRQSS` settings request, so you can query the
active protected attribute status.

## Validation Steps Performed

I've manually confirmed that we pass the selective erase tests in Vttest
now, and I've also manually tested some more complicated edge cases and
confirmed that we match the behavior of the hardware VT240 emulator in
MAME.

For unit testing I've extended the existing erase tests to cover
selective erase as an additional option, I've added a test covering the
`DECSCA` sequence, and I've extended the `DECRQSS` adapter test to
confirm the attribute reporting is working.

Closes #14029

### CODE REVIEW

File: src/terminal/adapter/adaptDispatch.cpp [hunk 1]
```diff
@@ -686,6 +686,103 @@ bool AdaptDispatch::EraseInLine(const DispatchTypes::EraseType eraseType)
...
+            for (auto col = eraseRect.left; col < eraseRect.right; col++)
+            {
+                // Only unprotected cells are affected.
+                if (!attrs.GetAttrByColumn(col).IsProtected())
+                {
+                    // The text is cleared but the attributes are left as is.
+                    chars.ClearGlyph(col);
+                    textBuffer.TriggerRedraw(Viewport::FromCoord({ col, row }));
+                }
+            }
```
Comment (j4james) [inline src/terminal/adapter/adaptDispatch.cpp:714]: I realise this bit will likely have to be rewritten somewhat to work with the new `ROW` implementation in PR #13626, so I'm happy to wait for that to merge first.

File: src/buffer/out/TextAttribute.cpp [hunk 2]
```diff
@@ -365,10 +375,11 @@ void TextAttribute::SetDefaultBackground() noexcept
 }
 
 // Method description:
-// - Resets only the rendition character attributes
+// - Resets only the rendition character attributes, which includes everything
+//     except the Protected attribute.
 void TextAttribute::SetDefaultRenditionAttributes() noexcept
 {
-    _attrs = CharacterAttributes::Normal;
+    _attrs &= CharacterAttributes::Protected;
```
Comment (DHowett) [inline src/buffer/out/TextAttribute.cpp:382]: Should we introduce a metaconstant for the rendition attributes? or should we just defer that until we have more non-rendition attributes... :)

File: src/terminal/adapter/adaptDispatch.cpp (hunk 1 above)
Comment (DHowett) [reply src/terminal/adapter/adaptDispatch.cpp:714]: `ROW` is going to take much longer than this PR to merge, so I'll ask Leonard to do that part :smile: Thanks though!

File: src/terminal/adapter/adaptDispatchGraphics.cpp [hunk 3]
```diff
@@ -259,6 +259,38 @@ bool AdaptDispatch::SetGraphicsRendition(const VTParameters options)
...
+        case Default:
+            attr.SetProtected(false);
+            break;
+        case Protected:
+            attr.SetProtected(true);
+            break;
+        case Unprotected:
+            attr.SetProtected(false);
+            break;
+        }
```
Comment (DHowett) [inline src/terminal/adapter/adaptDispatchGraphics.cpp:288]: Hmm. I like how clear this is, but it makes me wonder if it shouldn't be `attr.SetProtected(opt == Protected)` for some reason. I have no reason to request it specifically other than aesthetics, so f.f. to ignore :)

**Review by DHowett — APPROVED** [review]
Beautiful and straightforward code. Happy to merge this before `ROW`.

Noted: selective erase won't pass over ConPTY today--nor will protection--but due to the way ConPTY renders output it will technically work fine. Interesting!

**Comment by DHowett** [comment]
Also noted: part of making it transit ConPTY (if we were so inclined) would be to make triply sure the (ever-problematic) optimization to avoid sending empty spaces didn't optimize out a protection attribute :)

File: src/buffer/out/TextAttribute.cpp (hunk 2 above)
Comment (j4james) [reply src/buffer/out/TextAttribute.cpp:382]: I don't know. If we did, it would be everything but `Protected`, would seems like a pain to maintain. We could go the other way around and create a metaconstant for the "logical" attributes, but that would just be `Protected` which feels a bit pointless. I'd say defer it until have more non-rendition attributes, but realistically that'll probably be never.

**Comment by j4james** [comment]
> Also noted: part of making it transit ConPTY (if we were so inclined) would be to make triply sure the (ever-problematic) optimization to avoid sending empty spaces didn't optimize out a protection attribute :)

The protected attributes don't need to pass over conpty, because all of the logic is happening on the conhost side. Although now that I think about it, it's possible the existence of a protected attribute might prevent spaces being optimized where they potentially could have been, but that shouldn't break anything. I have tested this in Windows Terminal, and haven't seen any problems.

File: src/terminal/adapter/adaptDispatchGraphics.cpp (hunk 3 above)
Comment (DHowett) [reply src/terminal/adapter/adaptDispatchGraphics.cpp:288]: Ah, of course. Thanks!

File: src/terminal/adapter/adaptDispatch.cpp [hunk 4]
```diff
@@ -686,6 +686,103 @@ bool AdaptDispatch::EraseInLine(const DispatchTypes::EraseType eraseType)
...
+// Routine Description:
+// - Selectively erases unprotected cells in an area of the buffer.
+// Arguments:
+// - textBuffer - Target buffer to be erased.
+// - eraseRect - Area of the buffer that will be affected.
+// Return Value:
+// - <none>
+void AdaptDispatch::_SelectiveEraseRect(TextBuffer& textBuffer, const til::rect& eraseRect)
+{
+    if (eraseRect.left < eraseRect.right && eraseRect.top < eraseRect.bottom)
```
Comment (lhecker) [inline src/terminal/adapter/adaptDispatch.cpp:698]: You could just write `if (eraseRect)` here. The `operator bool` does exactly this, plus it ensures that all coordinates are >= 0, which seems to me like we might want in this code.

File: src/terminal/adapter/adaptDispatchGraphics.cpp (hunk 3 above)
Comment (lhecker) [inline src/terminal/adapter/adaptDispatchGraphics.cpp:288]: But what about this?
```cpp
switch (opt)
{
case Default:
case Unprotected: // <-- deduplicated
    attr.SetProtected(false);
    break;
case Protected:
    attr.SetProtected(true);
    break;
default: // <-- explicit (maybe?)
    break;
}
```

**Review by lhecker — COMMENTED** [review]
I'm ok with this PR. I'm just delaying my approval in case you wish to change anything. Otherwise I'll approve it anyways.

File: src/terminal/adapter/adaptDispatch.cpp (hunk 4 above)
Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:698]: Oh that's neat. I'll definitely do that.

File: src/terminal/adapter/adaptDispatchGraphics.cpp (hunk 3 above)
Comment (j4james) [reply src/terminal/adapter/adaptDispatchGraphics.cpp:288]: I did consider that, but the reason I left the `Default` separate from the `Unprotected` was because semantically they're actually different. `Default` serves the same purpose as `SGR 0` does for rendition attributes - it resets _all_ of the logical attributes. For now that just means resetting the protected attribute, because that's all there is, but if we ever added others, those two cases would definitely be different. I thought it better to get the structure right up front rather than trying to over optimise it.

File: src/terminal/adapter/adaptDispatchGraphics.cpp [hunk 5]
```diff
@@ -259,6 +259,38 @@ bool AdaptDispatch::SetGraphicsRendition(const VTParameters options)
...
+//   but the protected attribute was the only one ever implemented.
+// Arguments:
+// - options - An array of options that will be applied in order.
+// Return Value:
+// - True.
+bool AdaptDispatch::SetCharacterProtectionAttribute(const VTParameters options)
+{
+    auto& textBuffer = _api.GetTextBuffer();
+    auto attr = textBuffer.GetCurrentAttributes();
+    for (size_t i = 0; i < options.size(); i++)
```
Comment (christianparpart) [inline src/terminal/adapter/adaptDispatchGraphics.cpp:274]: Out of personal curiosity, when looking at the spec at https://vt100.net/docs/vt510-rm/DECSCA.html it looks like DECSCA accepts one single numeric parameter but you iterate over every one (from zero upwards?). Every subsequent parameter will override its predecessor. I can't find that in the linked spec. Why?

Comment (j4james) [reply src/terminal/adapter/adaptDispatchGraphics.cpp:274]: @christianparpart It's documented in DEC STD-070 on page 5-166 (I can get there directly in my copy of the PDF by selecting page 421). The sequence was designed to be extendable with additional attribute types similar to `SGR`. So just like you can have `\e[1;22m` turning bold on and off again, you can have `\e[1;2"q` turning the protected attribute on and off again.

And just like `\e[0m` turns off all rendition attributes, `\e[0"q` was actually intended to turn off all logical attributes. It's just that in this case that's effectively the same as `\e[2q`, because they never got around to defining any other logical attributes (at least as far as I'm aware).

Comment (christianparpart) [reply src/terminal/adapter/adaptDispatchGraphics.cpp:274]: Thanks for that clarification and reference.

**Comment by carlos-zamora** [comment]
I'm assuming this is ready to merge.

@msftbot merge this in 10 minutes
